dev/archery/archery/linking.py

```diff
@@ -61,9 +63,77 @@ def list_dependency_names(self):
             names.append(name)
         return names
 
+    def _extract_symbols(self, symbol_info):
+        return [re.search(r'\S+$', line).group() for line in symbol_info if line]
+
+    def _remove_weak_symbols(self, symbol_info):
+        return [line for line in symbol_info if not re.search(r'\s[Ww]\s', line)]
+
+    def _capture_symbols(self, remove_symbol_versions, symbol_info):
+        if remove_symbol_versions:
+            symbol_info = [re.split('@@', line)[0] for line in symbol_info]
+        return symbol_info
+
+    def list_symbols_for_dependency(self, dependency, remove_symbol_versions=False):
+        if dependency == 'linux-vdso.so.1':
+            return []
+        result = _nm.run('-D', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        lines = self._capture_symbols(remove_symbol_versions, lines)
+        return self._extract_symbols(lines)
+
+    def list_undefined_symbols_for_dependency(self, dependency,
+                                              remove_symbol_versions=False):
+        result = _nm.run('-u', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        lines = self._capture_symbols(remove_symbol_versions, lines)
+        lines = self._remove_weak_symbols(lines)
+        return self._extract_symbols(lines)
+
+    def extract_library_paths(self, file_path):
+        system = platform.system()
+        paths = {}
+        if system == 'Linux':
+            result = _ldd.run(file_path, stdout=subprocess.PIPE)
+            lines = result.stdout.decode('utf-8').splitlines()
+            for line in lines:
+                print(line)
+                match = re.search(r'(\S*) => (\S*)', line)
```

Comment:
Could you add an example result as a comment? For example:

```python
# Input:
#    librt.so.1 => /lib64/librt.so.1 (0x00007f46953f8000)
# Match:
#   group(1): librt.so.1
#   group(2): /lib64/librt.so.1
match = re.search(r'(\S*) => (\S*)', line)
```

Reply:
Yes, sure. I can do that.